huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go

```diff
@@ -0,0 +1,373 @@
+	}
+
+	resp, err := environments.Get(client, d.Id())
+	if err != nil {
+		common.CheckDeletedDiag(d, err, "error retrieving ServiceStage environment")
+	}
+
+	// Currently, the base resource type supports CCE, CCI, ECS and AS.
+	bTypes := []string{"cce", "cci", "ecs", "as"}
+	// Currently, the optional resource type supports ELB, EIP, RDS, DCS and CSE.
+	oTypes := []string{"elb", "eip", "rds", "dcs", "cse"}
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("name", resp.Name),
+		d.Set("description", resp.Description),
+		d.Set("vpc_id", resp.VpcId),
+	)
+	// Set if and only if the array of resources has at least one value to avoid 'terraform apply' triggering an update.
+	if val := flattenEnvironmentResources(resp.BaseResources, bTypes); len(val) > 0 {
+		mErr = multierror.Append(mErr, d.Set("basic_resources", val))
+	}
+	if val := flattenEnvironmentResources(resp.OptionalResources, oTypes); len(val) > 0 {
+		mErr = multierror.Append(mErr, d.Set("optional_resources", val))
+	}
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+// getResourcesChange is the method to get the change information of the specified resource type.
+func getResourcesChange(oldSet, newSet interface{}, rType string) ([]environments.Resource, []environments.Resource) {
+	addSet := newSet.(*schema.Set).Difference(oldSet.(*schema.Set))
+	addResult := make([]environments.Resource, addSet.Len())
+	for i, v := range addSet.List() {
+		addResult[i] = environments.Resource{
+			ID:   v.(string),
+			Type: rType,
+		}
+	}
+	rmSet := oldSet.(*schema.Set).Difference(newSet.(*schema.Set))
+	rmResult := make([]environments.Resource, rmSet.Len())
+	for i, v := range rmSet.List() {
+		rmResult[i] = environments.Resource{
+			ID:   v.(string),
+			Type: rType,
+		}
+	}
+
+	return addResult, rmResult
+}
+
+func getBaseResourcesChange(d *schema.ResourceData) (baseRes, rmRes []environments.Resource) {
+	if d.HasChange("basic_resources.0.cce_ids") {
+		oldSet, newSet := d.GetChange("basic_resources.0.cce_ids")
+		base, rm := getResourcesChange(oldSet, newSet, "cce")
+		baseRes = append(baseRes, base...)
+		rmRes = append(rmRes, rm...)
+	}
+	if d.HasChange("basic_resources.0.cci_ids") {
+		oldSet, newSet := d.GetChange("basic_resources.0.cci_ids")
+		base, rm := getResourcesChange(oldSet, newSet, "cci")
+		baseRes = append(baseRes, base...)
+		rmRes = append(rmRes, rm...)
+	}
+	if d.HasChange("basic_resources.0.ecs_ids") {
+		oldSet, newSet := d.GetChange("basic_resources.0.ecs_ids")
+		base, rm := getResourcesChange(oldSet, newSet, "ecs")
+		baseRes = append(baseRes, base...)
+		rmRes = append(rmRes, rm...)
+	}
+	if d.HasChange("basic_resources.0.as_ids") {
+		oldSet, newSet := d.GetChange("basic_resources.0.as_ids")
+		base, rm := getResourcesChange(oldSet, newSet, "as")
+		baseRes = append(baseRes, base...)
+		rmRes = append(rmRes, rm...)
+	}
+	return
+}
+
+func getOptionalResourcesChange(d *schema.ResourceData) (optRes, rmRes []environments.Resource) {
```

Comment:
please consider merge this func with getBaseResourcesChange

Reply:
done.